src/puppetlabs/puppetdb/scf/migrate.clj

```diff
@@ -1546,16 +1546,31 @@
        new_value, old_value, message, file, line, containment_path, containing_class, corrective_change
      from resource_events"]
    (fn [rows]
-     (reduce (fn [hashes-seen row]
-               (conj hashes-seen
-                     (let [hash-str (hash/resource-event-identity-pkey row)]
-                       (when-not (hashes-seen hash-str)
-                         (jdbc/insert-multi! "resource_events_transform"
-                                             (list (assoc row :event_hash
-                                                          (sutils/munge-hash-for-storage hash-str)))))
-                       hash-str)))
-             #{}
-             rows)))
+     (let [old-cols '(:report_id :certname_id :status :timestamp :resource_type
+                      :resource_title :property :new_value :old_value :message
+                      :file :line :containment_path :containing_class
+                      :corrective_change)
+           new-cols (conj old-cols :event_hash)
+           update-row (apply juxt (conj old-cols
+                                        (comp sutils/munge-hash-for-storage
+                                              hash/resource-event-identity-pkey)))
+           insert->hash (fn [batch]
+                          (jdbc/insert-multi! :resource_events_transform
+                                              new-cols
+                                              batch)
+                          ;; Return hashes for each row directly instead of
+                          ;; reading them back from the DB via the return
+                          ;; value of insert-multi!
+                          (map :event_hash batch))
+           dedupe-and-insert (fn [hashes-seen batch]
+                               (->> batch
+                                    (pmap update-row)
```

Comment:
@rbrw You mentioned there may be some issues with using `pmap` vs `map`. It's a minor optimization compared to the boost we get from batching, but hashing is CPU-intensive enough that it seems worth doing in parallel if we can do so easily.

Reply:
For the test dataset of 5.7 million rows, `pmap` appeared to cut about 3.5 minutes off of a 15 minute batched migration.

Reply:
Yeah, let's check.  I think the concern may have been the fact that pmap is a bit blunt, and uses the common agent pool.  If that's it, then ideally you'd want to parallelize via a separate pool, though it might be less of a concern during startup.  I think part of that (serious) general concern is that in general you might end up with all the threads blocked (affecting agents, core.async, etc. too), and possibly even deadlocking the program?

Reply:
Removed `pmap`.